File: marcov/marcov-csv.rb
#!/usr/bin/env ruby
#-*- coding:utf-8 -*-
require 'natto'

def wakati(text)
  words = []
  nm = Natto::MeCab.new
  nm.parse(text) do |n|
    words << n.surface
  end
  return words
end

def markov_d2(src)
#  wordlist = src
  wordlist = wakati(src)
  markov = {}
  w1 = ""
#  wordlist.split(//).each do |word|
  wordlist.each do |word|
    if w1
      unless markov.include?(w1)
        markov[w1] = []
      end
      markov[w1] << word
    end
    w1=word
  end
  count=0
  sentence=""
  w1=markov.keys.sample
  while count<20
    if markov.has_key?(w1)
      tmp = markov[w1].sample
      sentence += tmp
    end
    w1=tmp
    count+=1
  end
  return sentence
end

def markov_d3(src)
#  wordlist = src
  wordlist = wakati(src)
  markov = {}
  w1 = ""
  w2 = ""
#  wordlist.split(//).each do |word|
  wordlist.each do |word|
    if w1 && w2
      unless markov.include?([w1,w2])
        markov[[w1,w2]] = []
      end
      markov[[w1,w2]] << word
    end
    w1,w2=w2,word
  end
  count=0
  sentence=""
  w1,w2=markov.keys.sample
  while count<20
    if markov.has_key?([w1,w2])
      tmp = markov[[w1,w2]].sample
      sentence += tmp
    end
    w1,w2=w2,tmp
    count+=1
  end
  return sentence
end

def markov_d4(src)
  wordlist = src
#  wordlist = wakati(src)
  markov = {}
  w1 = ""
  w2 = ""
  w3 = ""
  wordlist.split(//).each do |word|
#  wordlist.each do |word|
    if w1 && w2 && w3
      unless markov.include?([w1,w2,w3])
        markov[[w1,w2,w3]] = []
      end
      markov[[w1,w2,w3]] << word
    end
    w1,w2,w3=w2,w3,word
  end
  count=0
  sentence=""
  w1,w2,w3=markov.keys.sample
  while count<20
    if markov.has_key?([w1,w2,w3])
      tmp = markov[[w1,w2,w3]].sample
      sentence += tmp
    end
    w1,w2,w3=w2,w3,tmp
    count+=1
  end
  return sentence
end

dir = "./d2c_201410-11.txt"
dir = "../nlp/nippo/nippos/tech/07/all.txt"

text = open((dir), "r").readlines.uniq.join("").gsub(/\n|\"/,"")
30.times do |i|
  p i
  open("markov.txt","a+") do |f|
#  open("google_parse_d2_res.txt","a+") do |f|
    f.puts markov_d2(text)
  end
end

File: marcov/marcov-csv.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import MeCab
import sys
import random
import re
import pprint
import yaml
import pandas as pd

def pp(obj):
  pp = pprint.PrettyPrinter(indent=4, width=160)
  str = pp.pformat(obj)
  return re.sub(r"\\u([0-9a-f]{4})", lambda x: unichr(int("0x"+x.group(1), 16)), str)

def wakati(text):
  t = MeCab.Tagger("-Owakati")
  m = t.parse(text)
  result = m.rstrip(" \n").split(" ")
  return result

def markov_d1(src):
  wordlist = src#wakati(src)
  markov = {}
  w1=''
  for word in wordlist:
    if w1:
      if(w1)not in markov:
        markov[(w1)] = []
      markov[(w1)].append(word)
    w1=word
  count = 0
  sentence=''
  w1=random.choice(markov.keys())
  #出力description数
  while count < 20:
    if markov.has_key((w1))==True:
      tmp = random.choice(markov[(w1)])
      sentence += tmp
    w1=tmp
    count += 1
  return sentence

def markov_d2(src):
  wordlist = src#wakati(src)
  markov = {}
  w1=''
  w2=''
  for word in wordlist:
    if w1 and w2:
      if(w1, w2)not in markov:
        markov[(w1, w2)] = []
      markov[(w1, w2)].append(word)
    w1,w2=w2,word
  count = 0
  sentence=''
  w1,w2=random.choice(markov.keys())
  #出力description数
  while count < 20:
    if markov.has_key((w1,w2))==True:
      tmp = random.choice(markov[(w1,w2)])
      sentence += tmp
    w1=tmp
    count += 1
  return sentence


def markov_d3(src):
  wordlist = src#wakati(src)
  markov = {}
  w1=''
  w2=''
  w3=''
  for word in wordlist:
    if w1 and w2 and w3:
      if(w1, w2, w3)not in markov:
        markov[(w1, w2, w3)] = []
      markov[(w1, w2, w3)].append(word)
    w1,w2,w3=w2,w3,word
  count = 0
  sentence=''
  w1,w2,w3=random.choice(markov.keys())
  #出力description数
  while count < 20:
    if markov.has_key((w1,w2,w3))==True:
      tmp = random.choice(markov[(w1,w2,w3)])
      sentence += tmp
    w1=tmp
    count += 1
  return sentence


#df = pd.read_csv("./d2c_201410-11.csv", encoding="cp932")
#print df.values
#texts = []
#f = open("./d2c_201410-11.csv", "rb")
#texts = texts.fromfile(f)
#print texts

with open("./d2c_201410-11.txt", "rb") as f:
  texts = f.readlines()
texts.pop(0)

#print(pp(texts[0]))
print(len(texts))
print(len(wakati("".join(texts))))

#print(markov_d2(texts))
